### PR TITLE
NeuralMAB

### PR DESCRIPTION
**Pull request by maypink**
- Implementation of Contextual Bandits and Neural Contextual Bandits from the [article](https://arxiv.org/abs/2012.01780). 
- SyntheticExperimentHelper to compare bandits with each other

### CODE REVIEW

File: golem/core/optimisers/adaptive/neural_mab.py [hunk 1]
```diff
@@ -0,0 +1,306 @@
...
+    Deep representation is formed with NN and Contextual Multi-Armed Bandit is integrated to choose arm.
+    """
+    def __init__(self, arms: List[Arm],
+                 learning_policy: Any = LearningPolicy.UCB1(alpha=0.5),
+                 neighborhood_policy: Any = NeighborhoodPolicy.TreeBandit(),
+                 seed: int = Constants.default_seed,
+                 n_jobs: int = 1):
+
+        super().__init__(arms, learning_policy, neighborhood_policy, seed, n_jobs)
+        self.nn_with_se = NNWithShallowExploration(context_size=500, arms_count=len(arms))
```
Comment (gkirgizov) [inline golem/core/optimisers/adaptive/neural_mab.py:30]: А здешний `context_size` ты тоже обрезала до 20 при попытке сократить время? Вижу только двадцатку выше, в энкодере

File: golem/core/optimisers/adaptive/neural_mab.py [hunk 2]
```diff
@@ -0,0 +1,306 @@
...
+
+    def _initial_fit_mab(self, context: Any):
+        """ Initial fit for Contextual Multi-Armed Bandit.
+        At this step, all hands are assigned the same weights with the very first context
+        that is fed to the bandit. """
+        # initial fit for mab
+        n = len(self.arms)
+        uniform_rewards = [1. / n] * n
+        deep_context = self._get_deep_context(context=context)
+        self._mab.fit(decisions=self._indices, rewards=uniform_rewards, contexts=n * [deep_context])
```
Comment (gkirgizov) [inline golem/core/optimisers/adaptive/neural_mab.py:52]: Предлагаю убрать этот обман бедного бандита с произвольным контекстом. Вместо этого использовать флаг is_fitted внутри partial_fit для первого фита

File: golem/core/optimisers/adaptive/neural_mab.py [hunk 3]
```diff
@@ -0,0 +1,306 @@
...
+        self.THETA_action = torch.tensor([])
+        self.CONTEXT_action = torch.tensor([])
+        self.REWARD_action = torch.tensor([])
+        self.result_neuralucb = []
+        self.W = copy.deepcopy(self.W0)
+        self.summ = 0
+        self.log = default_log('NNWithShallowExploration')
+
+    def partial_fit(self, iter: int,
+                    decisions: List[Any], rewards: List[float], contexts: List[Any] = None):
```
Comment (gkirgizov) [inline golem/core/optimisers/adaptive/neural_mab.py:121]: Можешь пояснить этот кусок? Мы обучаемся раз в H итераций оптимизатора? Мы каждую итерацию что-то делаем?

File: golem/core/optimisers/adaptive/neural_mab.py [hunk 4]
```diff
@@ -0,0 +1,306 @@
...
+        return self._mab.predict_expectations(contexts=[deep_context])
+
+    def _get_deep_context(self, context: Any) -> List[Any]:
+        """ Returns deep representation of context. """
+        temp = self.nn_with_se.transfer(context, 0, len(self.arms))
+        feat = self.nn_with_se.feature_extractor(temp, self.nn_with_se.W).numpy().squeeze()
+        return list(feat)
+
+
+class NNWithShallowExploration:
```
Comment (gkirgizov) [inline golem/core/optimisers/adaptive/neural_mab.py:86]: Это твоя реализация или она взята откуда-то?

**Review by gkirgizov — CHANGES_REQUESTED** [review]
Все выглядит разумно, вопросов по реализации нет, есть вопросы и комментарии по логике.

- Сориентируй, как долго работает энкодер на графах? Сколько занимает итерация обучения нейронки?
- Тестировала ли ты реализацию бандита на какой-то простой задаче? Бенчмарк из статьи, например. Можно без энкодера графов обойтись.
- Нужно поставить эксперимент, где для разных графов (разных эмбеддингов) наилучшими будут различные действия. Например, если это оптимизация размера графов до N , то маленькие графы надо растить, а большие обрезать. Вот надо бы хотя бы такой тривиальный случай поймать, чтобы проверить , что контекст вообще работает. И затем что-то менее тривиальное надо придумать.

File: golem/core/adapter/nx_adapter.py [hunk 5]
```diff
@@ -85,3 +85,24 @@ def _node_restore(self, node: GraphNode) -> Dict:
 
     def _node_adapt(self, data: Dict) -> OptNode:
         return data[_NX_NODE_KEY]
+
+
+class BanditNetworkxAdapter(BaseNetworkxAdapter):
```
[the remark this answers is not in the document]
Comment (maypink) [reply golem/core/adapter/nx_adapter.py:90]: Потому что с обычным адаптером там в качестве имен нод в nx.Graph определеляется uid из големовского графа =>в karateclub(откуда импортируется класс для получения эмбеддингов) есть проверка, которая не проходит из-за этого
image

File: golem/core/optimisers/adaptive/neural_mab.py (hunk 1 above)
Comment (maypink) [reply golem/core/optimisers/adaptive/neural_mab.py:30]: не, обрезала везде, там же иначе просто матрицы не перемножатся из-за несовпадения размерностей. это я скорее 20 забыла убрать

File: golem/core/optimisers/adaptive/neural_mab.py (hunk 3 above)
Comment (maypink) [reply golem/core/optimisers/adaptive/neural_mab.py:121]: обучаемся и обновляем веса раз в _H_q (сейчас 5) итераций, но каждую итерацию подсчитываем текущий reward и собираем датасет для следующего дообучения (в context_action и reward_action)

File: golem/core/optimisers/adaptive/neural_mab.py (hunk 4 above)
Comment (maypink) [reply golem/core/optimisers/adaptive/neural_mab.py:86]: нейронка взята с supplementary материалов к этой статье, с open review

File: golem/core/optimisers/adaptive/neural_contextual_mab_agent.py [hunk 6]
```diff
@@ -0,0 +1,59 @@
...
+    def partial_fit(self, experience: ExperienceBuffer):
+        """Continues learning of underlying agent with new experience."""
+        obs, actions, rewards = experience.retrieve_experience()
+        self._dbg_log(obs, actions, rewards)
+        arms = [self._arm_by_action[action] for action in actions]
+        obs_embs = self._get_obs_embedding(obs=obs)
+        self._agent.partial_fit(decisions=arms, rewards=rewards, contexts=obs_embs)
+
+    def _get_obs_embedding(self, obs: List[ObsType]) -> List[Any]:
+        context_agent = ContextAgentsRepository.agent_class_by_id(self._context_agent)
```
[the remark this answers is not in the document]
Comment (maypink) [reply golem/core/optimisers/adaptive/neural_contextual_mab_agent.py]: [done](https://github.com/aimclub/GOLEM/pull/102/commits/9893a281cfb8f2ae9ad06f866f3a7ecdb710be75#diff-ce10550c5a8aaaf09d20512132eab2b49192b96b1072a68e0a3b088c4635abd4R23)

File: golem/core/optimisers/adaptive/neural_mab.py (hunk 2 above)
Comment (maypink) [reply golem/core/optimisers/adaptive/neural_mab.py:52]: не совсем поняла, при чем тут первый фит. проблема же в том, что predict вызывается первее, чем partial_fit, так что нужен какой-то warm start

File: golem/core/adapter/nx_adapter.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (gkirgizov) [reply golem/core/adapter/nx_adapter.py:90]: И зачем им это нужно... 
Добавь где-нибудь комментарий с обоснованием, пожалуйста

File: golem/core/optimisers/adaptive/neural_mab.py (hunk 2 above)
Comment (gkirgizov) [reply golem/core/optimisers/adaptive/neural_mab.py:52]: может, тогда в predict добавить проверку на is_fitted? и выдавать дефолтные вероятности, что-то такое

File: golem/core/optimisers/adaptive/neural_mab.py (hunk 3 above)
Comment (gkirgizov) [reply golem/core/optimisers/adaptive/neural_mab.py:122]: Понял, спасибо. Добавь, пожалуйста, это где-нибудь в комментарий. Позволяет быстрее разобраться в логике

File: golem/core/optimisers/adaptive/neural_mab.py (hunk 4 above)
Comment (gkirgizov) [reply golem/core/optimisers/adaptive/neural_mab.py:87]: оо, понял о чем ты говорила про кучу кода в статье. Нет бы им по-человечески залить на гитхаб. Хитрые ребята в  этом DeepMind, явно что-то скрывают

File: golem/core/optimisers/adaptive/neural_mab.py (hunk 2 above)
Comment (maypink) [reply golem/core/optimisers/adaptive/neural_mab.py:52]: да будто бы это оно и есть. мы же одинаковый контекст с одинаковыми наградами передаем для всех рук, то есть вероятности не сдвинутся после такого

File: golem/core/adapter/nx_adapter.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (maypink) [reply golem/core/adapter/nx_adapter.py:89]: [done](https://github.com/aimclub/GOLEM/pull/102/commits/22d3305886c1fbcd6573cb11ae51860ab33cce81#diff-78c47341b9a2c95965f9d7f4c138d331ac0ae6b60c8934fbfeb447e20c836433R90)

File: golem/core/optimisers/adaptive/neural_mab.py (hunk 3 above)
Comment (maypink) [reply golem/core/optimisers/adaptive/neural_mab.py:122]: [done](https://github.com/aimclub/GOLEM/pull/102/commits/9893a281cfb8f2ae9ad06f866f3a7ecdb710be75#diff-58861d15f7642ee0468b2f3423ffdc145f0af1f3c10e7fdd5f093375990f732cR57)

File: golem/core/optimisers/adaptive/neural_mab.py (hunk 4 above)
Comment (maypink) [reply golem/core/optimisers/adaptive/neural_mab.py:87]: да спасибо, что хоть так выложили

File: golem/core/optimisers/adaptive/neural_mab.py (hunk 2 above)
Comment (gkirgizov) [reply golem/core/optimisers/adaptive/neural_mab.py:53]: Мы вносим bias , давая бандиту ложные данные. Это не очень корректно, больше на хак похоже.

File: examples/adaptive_optimizer/experiment_setup.py [hunk 7]
```diff
@@ -47,3 +49,48 @@ def log_action_values(next_pop: PopulationT, optimizer: EvoGraphOptimizer):
...
+    cluster = KMeans(n_clusters=n_clusters)
+
+    def log_action_values_with_clusters(next_pop: PopulationT, optimizer: EvoGraphOptimizer):
+        obs_contexts = optimizer.mutation.agent.get_context(next_pop)
+        cluster.fit(np.array(obs_contexts).reshape(-1, 1))
+        centers = cluster.cluster_centers_
+        for i, center in enumerate(centers):
+            values = optimizer.mutation.agent.get_action_values(obs=center)
+            if i not in stats_action_value_log.keys():
+                stats_action_value_log[i] = (list(values))
```
Comment (gkirgizov) [inline examples/adaptive_optimizer/experiment_setup.py:73]: Классная мысль использовать KMeans для этого

File: examples/adaptive_optimizer/mab_experiment_different_targets.py [hunk 8]
```diff
@@ -68,6 +69,7 @@ def run_experiment_node_num(adaptive_mutation_type: MutationAgentTypeEnum,
                                                      adaptive_mutation_type=adaptive_mutation_type),
             timeout=timedelta(minutes=trial_timeout),
             num_iterations=target_size * 3,
+            graph_size=[random.randint(5, 10) for _ in range(19)] + [random.randint(90, 95) for _ in range(2)]
```
Comment (gkirgizov) [inline examples/adaptive_optimizer/mab_experiment_different_targets.py:72]: Предлгааю тут разделить поровну общий pop_size

File: examples/synthetic_graph_evolution/graph_search.py [hunk 9]
```diff
@@ -25,7 +25,9 @@ def graph_search_setup(target_graph: Optional[nx.DiGraph] = None,
                        algorithm_parameters: Optional[AlgorithmParameters] = None,
                        node_types: Sequence[str] = ('x',),
                        timeout: Optional[timedelta] = None,
-                       num_iterations: Optional[int] = None):
+                       num_iterations: Optional[int] = None,
+                       graph_size: Optional[List[int]] = None,
```
Comment (gkirgizov) [inline examples/synthetic_graph_evolution/graph_search.py:29]: Предлагаю переименовать в `initial_graph_sizes` для очевидности

File: experiments/mab/mab_synthetic_experiment_helper.py [hunk 10]
```diff
@@ -0,0 +1,206 @@
...
+                values = optimizer.mutation.agent.get_action_values(obs=[center])
+                if i not in stats_action_value_log.keys():
+                    stats_action_value_log[i] = []
+                stats_action_value_log[i].append(list(values))
+
+        # set iteration callback
+        if bandit_type == MutationAgentTypeEnum.bandit:
+            optimizer.set_iteration_callback(log_action_values)
+        else:
+            optimizer.set_iteration_callback(log_action_values_with_clusters)
```
Comment (gkirgizov) [inline experiments/mab/mab_synthetic_experiment_helper.py:85]: Добавь лучше явное условие. А то при запуске, например, на `default` vs `bandit` получим что-то не то

File: experiments/mab/mab_synthetic_experiment_helper.py [hunk 11]
```diff
@@ -0,0 +1,206 @@
...
+            self.bandit_metrics[bandit_type.name] = []
+        self.bandit_metrics[bandit_type.name].append(final_metrics)
+
+        print('History of action probabilities:')
+        pprint(stats_action_value_log)
+        if self.is_visualize:
+            self.show_fitness_line(found_nx_graph=found_nx_graph, final_metrics=final_metrics,
+                                   history=history)
+            self.show_action_probabilities(bandit_type=bandit_type, stats_action_value_log=stats_action_value_log,
+                                           actions=agent.actions)
```
Comment (gkirgizov) [inline experiments/mab/mab_synthetic_experiment_helper.py:106]: Дело вкуса, пожалуй, но я вижу роль kwargs в 3 случаях: когда мы указываем доп. параметры, когда мы меняем порядок аргументов по каким-то своим причинам, или когда это улучшает читаемость (например, `range(start=42, end=99)`). В данном случае, по мне, как-то избыточно

File: experiments/mab/mab_synthetic_experiment_helper.py [hunk 12]
```diff
@@ -0,0 +1,206 @@
...
+        for bandit in list(show_action_probabilities.keys()):
+            total_sum = None
+            for launch in show_action_probabilities[bandit]:
+                if not total_sum:
+                    total_sum = launch
+                    continue
+                for cluster in launch.keys():
+                    for i in range(len(total_sum[cluster])):
+                        for j in range(len(total_sum[cluster][i])):
+                            total_sum[cluster][i][j] += launch[cluster][i][j]
```
Comment (gkirgizov) [inline experiments/mab/mab_synthetic_experiment_helper.py:143]: о, впервые вижу вложенный цикл глубины 5 )
(этот комент действий не требует)

File: requirements.txt [hunk 13]
```diff
@@ -4,9 +4,9 @@ pandas>=1.3.0; python_version >='3.8'
 
 # Models and frameworks
 networkx>=2.4, !=2.7.*, !=2.8.1, !=2.8.2, !=2.8.3
-scipy
-zss
-mabwiser  # for adaptive learning
+scipy~=1.7.3
+zss~=1.2.0
+mabwiser~=2.7.0
```
Comment (gkirgizov) [inline requirements.txt:9]: Перенеси этого mabwiser к остальным зависимостям бандитов и, пожалуй, вынеси эти зависимости вместе с PyTorch в какой-нибудь отдельный requirements-adaptive. Это пока экспериментальная функциональность и не хочется ее зависимостями утяжелять ГОЛЕМ. Тем более только избавились в дочернем ФЕДОТе от тяжелых зависимостей в основном файле.

**Review by gkirgizov — CHANGES_REQUESTED** [review]

File: experiments/mab/mab_synthetic_experiment_helper.py (hunk 11 above)
Comment (maypink) [reply experiments/mab/mab_synthetic_experiment_helper.py:106]: у меня просто было пару случаев, когда аргументы путались между собой из-за отсутствия четкого сопоставления, поэтому я все-таки за однозначную определенность здесь

File: examples/adaptive_optimizer/mab_experiment_different_targets.py (hunk 8 above)
Comment (maypink) [reply examples/adaptive_optimizer/mab_experiment_different_targets.py]: [done](https://github.com/aimclub/GOLEM/pull/102/commits/6446b72435833b425563dfff5cb2cc7656b22a2d#diff-7604b6987015309112ebfa7a33971f563b35424111e458867f4dee965c9a7417L72), убрала вообще, а то там твоя постановка задачи была, а я сверху поэкспериментировала

File: examples/synthetic_graph_evolution/graph_search.py (hunk 9 above)
Comment (maypink) [reply examples/synthetic_graph_evolution/graph_search.py]: [done](https://github.com/aimclub/GOLEM/pull/102/commits/6446b72435833b425563dfff5cb2cc7656b22a2d#diff-689a685e451a5f79c85eb2586f23af951f3ac75335d4b6aae2e5408a5619c339R29)

File: experiments/mab/mab_synthetic_experiment_helper.py (hunk 10 above)
Comment (maypink) [reply experiments/mab/mab_synthetic_experiment_helper.py]: [done](https://github.com/aimclub/GOLEM/pull/102/commits/6446b72435833b425563dfff5cb2cc7656b22a2d#diff-b6cf9baf15bb8280215954f2c5c9321bacb8f545dc267a54f81737e75bddc17eR84)

File: requirements.txt (hunk 13 above)
Comment (maypink) [reply requirements.txt]: [done](https://github.com/aimclub/GOLEM/pull/102/commits/6446b72435833b425563dfff5cb2cc7656b22a2d#diff-5f60f4ecdeaa9f75e989fa3070db7af8805dbde78d7386e6259467d96127a185R1)

File: golem/core/optimisers/adaptive/neural_mab.py (hunk 2 above)
Comment (maypink) [reply golem/core/optimisers/adaptive/neural_mab.py:53]: да вроде не можем: у бандитов есть warm_start, но он предназначен не для этого, а для того, чтобы просто задать начальное представление перед первым фитом. без фита все равно не обойтись

**Review by gkirgizov — APPROVED** [review]
